### ISSUE TITLE
vth is too large for some models

### ISSUE THREAD

**Issue by tcstewar** [nengo/nengo-loihi#59]
When running some of the benchmark models, I'm getting this error fairly frequently:

```
File "/home/terry/nengo-loihi/nengo_loihi/loihi_api.py", line 140, in validate
    core.validate()
  File "/home/terry/nengo-loihi/nengo_loihi/loihi_api.py", line 193, in validate
    vthProfile.validate(core=self)
  File "/home/terry/nengo-loihi/nengo_loihi/loihi_api.py", line 332, in validate
    assert 0 < self.vth <= VTH_MAN_MAX
```

It seems to be dependent on the random number seed -- adjusting the seed can make the issue go away.   The simplest model I've been able to build which replicates this is here:

```python
import nengo
import nengo.spa as spa
import nengo_loihi
import numpy as np

np.random.seed(2)

model = spa.SPA()
with model:
    n = nengo.Node(None, size_in=1)

    vocab = spa.Vocabulary(1)
    vocab.parse('A+B')
    model.cleanup = spa.AssociativeMemory(vocab)

    nengo.Connection(n, model.cleanup.input)
    probe = nengo.Probe(model.cleanup.output, synapse=0)

with nengo_loihi.Simulator(model, precompute=False) as sim:
    sim.run(0.1)
```

Interestingly, if I do any of the following, I can't get it to fail (after trying 10 different random number seeds):
 - getting rid of the connection from the node to the cleanup
 - reducing the parse to ('A') (i.e. only cleaning up to one item)
 - changing the Probe synapse to None
 - switching to precompute=True

So I'm not sure what's happening here, but I'm hoping someone will have some insights...  :)

**Comment by hunse** [comment]
Ok! I was curious why this only happens on the chip, since I'm pretty sure I checked this earlier in the build process.

Turns out, the problem is not that the `vth` is too large; it's that it's too small! It's the `< 0` part of the assertion that's tripping. There's a similar check in `vth_to_manexp`, but it only checks that `vth >= 0`, so it wasn't catching this.

I'm looking more into why `vth` ends up being zero, but the first thing I've noticed is that this model has some very large weights (the max weight is around `1e17`). This seems to be throwing `discretize` for a loop. Any ideas why the weights might be so large, @tcstewar? I feel like we're going to have trouble representing such large weights no matter what we do in `discretize`.

**Comment by tcstewar** [comment]
> So it appears in `AssociativeMemory.am_ens_config`, the intercepts range from zero to one. This creates some neurons with almost-one intercepts, and thus very high gains, resulting in large values in the scaled encoders.

Hmm...  This seems to me like it would be a problem for almost *all* nengo models, since the normal default intercept range is -1 to 1.  This particular case of setting it to an intercept of 0 to 1 should only double the chance of any particular neuron having too high a gain.  Am I interpreting this right?

> For now, we might just have to see if there's a way we can override the default intercepts and max rates for SPA components, to make them suitable for Loihi.

If I'm right about it also being a problem with intercepts between -1 and 1, then it feels to me like this isn't a SPA-specific problem.  Instead, I'd be more tempted to do this in the builder itself.  If we can find out some sort of upper bound on a reasonable intercept for loihi, then we could round any intercept higher than that down to that value.  That seems consistent with the general idea that backends are free to adjust what is being requested of them based on the hardware available (for example, even though we as nengo_loihi for a LIF neuron, we don't get exactly an LIF neuron, and even though we ask for particular connection weights, we don't get exactly that connection weight due to discretization).

**Comment by hunse** [comment]
I like that by handling it in the builder, the user doesn't have to worry
about it at all. One concern, though, is that rounding all the large
intercepts down will leave a lot of neurons with identical intercepts at
whatever our upper bound is; not the most efficient. Another concern is
that it is less transparent, though at least our current system should
allow the user to get the biases back from the builder and see what they've
actually been set to.

Maybe we do the rounding combined with a warning, so that users can be
aware and fix the intercepts distribution in their model if they so wish.

**Comment by tcstewar** [comment]
> Maybe we do the rounding combined with a warning, so that users can be aware and fix the intercepts distribution in their model if they so wish.

Sounds good to me!  :)

**Comment by drasmuss** [comment]
Fixed in #69
